tests/MongoDB.Driver.Core.Tests/Specifications/connection-monitoring-and-pooling/ConnectionMonitoringAndPoolingTestRunner.cs

```diff
@@ -0,0 +1,609 @@
+                            break;
+                        default:
+                            throw new ArgumentException($"Unknown pool option {poolOption.Name}.");
+                    }
+                }
+            }
+        }
+
+        public void ResetConnectionId()
+        {
+            IdGeneratorReflector.__lastId(0);
+        }
+
+        private IConnectionPool SetupConnectionPool(BsonDocument test, IEventSubscriber eventSubscriber)
+        {
+            var endPoint = new DnsEndPoint("localhost", 27017);
+            var serverId = new ServerId(new ClusterId(), endPoint);
+            ParseSettings(test, out var connectionPoolSettings, out var connectionSettings);
+
+            var connectionFactory = new Mock<IConnectionFactory>();
+            connectionFactory
+                .Setup(c => c.CreateConnection(serverId, endPoint))
+                .Returns(() =>
+                {
+                    var connection = new MockConnection(serverId, connectionSettings, eventSubscriber);
+                    connection.Open(CancellationToken.None);
+                    return connection;
+                });
+            var connectionPool = new ExclusiveConnectionPool(
+                serverId,
+                endPoint,
+                connectionPoolSettings,
+                connectionFactory.Object,
+                eventSubscriber);
+
+            return connectionPool;
+        }
+
+        private void Start(BsonDocument operation, ConcurrentDictionary<string, Task> tasks)
+        {
+            var startTarget = operation.GetValue("target").ToString();
+            tasks.GetOrAdd(startTarget, (Task)null);
+        }
+
+        private void WaitForEvent(EventCapturer eventCapturer, BsonDocument operation)
+        {
+            var eventType = MapExpectedEventName(operation.GetValue("event").ToString());
+            if (eventType == null)
+            {
+                return;
+            }
+
+            var expectedCount = operation.GetValue("count").ToInt32();
+            var notifier = new EventNotifier(
+                eventCapturer,
+                coll => coll.Count(c => c.GetType().Name == eventType) >= expectedCount);
+
+            var testFailedTimeout = Task.Delay(TimeSpan.FromMinutes(5), CancellationToken.None);
+            var index = Task.WaitAny(notifier.TaskCompletion, testFailedTimeout);
+            if (index != 0)
+            {
+                throw new Exception($"{nameof(WaitForEvent)} executing is too long.");
+            }
+        }
+
+        private void WaitForThread(BsonDocument operation, ConcurrentDictionary<string, Task> tasks, out Exception exception)
+        {
+            exception = null;
+            var waitThread = operation.GetValue("target").ToString();
+            if (tasks.TryGetValue(waitThread, out var task) && task != null)
+            {
+                exception = Record.Exception(() => task.GetAwaiter().GetResult());
```

Comment:
I use `GetAwaiter().GetResult()` intead of `Wait()` because `Wait()` packs errors in `AggregateException`